### PR TITLE
add hgu cache

### PR DESCRIPTION
**Pull request by xzav3r**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: libs/cosmos-sdk/baseapp/gasuseddb.go [hunk 1]
```diff
@@ -19,15 +22,107 @@ const (
...
+	if !cacheHit && hgu != -1 {
+		// add to cache before returning hgu
+		h.cache.Add(string(key), hgu)
+	}
+	totalCount++
+	return hgu
+}
+
+func (h *HistoryGasUsedRecordDB) FlushHgu() {
+	fmt.Println("debug", totalCount, hitCount)
```
Comment (zjg555543) [inline libs/cosmos-sdk/baseapp/gasuseddb.go]: ignore this statement?

Comment (xzav3r) [reply libs/cosmos-sdk/baseapp/gasuseddb.go]: > ignore this statement?

fixed

File: libs/cosmos-sdk/baseapp/gasuseddb.go [hunk 2]
```diff
@@ -19,15 +21,102 @@ const (
...
+	hgu, cacheHit := h.getHgu(key)
+	if !cacheHit && hgu != -1 {
+		// add to cache before returning hgu
+		h.cache.Add(string(key), hgu)
+	}
+	return hgu
+}
+
+func (h *HistoryGasUsedRecordDB) FlushHgu() {
+	if len(h.latestGu) == 0 {
```
Comment (zjg555543) [inline libs/cosmos-sdk/baseapp/gasuseddb.go:75]: Does it need lock by latestGuMtx？

Comment (zjg555543) [reply libs/cosmos-sdk/baseapp/gasuseddb.go:75]: it's a copy and delete after endbock

**Review by zjg555543 — APPROVED** [review]

**Pull request merged**